### PR TITLE
[MM-22893] Make UserController type to be used configurable

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR will make the type of `UserController` to run during a load-test a configurable option. It also adds a configuration setting for the rate.

#### Ticket

https://mattermost.atlassian.net/browse/MM-22893

### CODE REVIEW

File: cmd/loadtest/loadtest.go [hunk 1]
```diff
@@ -22,7 +23,15 @@ func RunLoadTestCmdF(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
-	newSimpleController := func(id int, status chan<- control.UserStatus) control.UserController {
+	if ok, err := config.IsValid(); !ok {
```
Comment (isacikgoz) [inline cmd/loadtest/loadtest.go]:
```suggestion
	if _, err := config.IsValid(); err != nil {
```

If we are going to return `error` (which is always false if error) simply checking `err` would be more idiomatic. That said, do we need `bool, error` return type for `func (c *Config) IsValid() (bool, error) `? Don't think so, it never returns `true, error` or `false, nil`. See [implementation](https://github.com/streamer45/mattermost-load-test-ng/blob/ae82536590d6956f2c75b4bea8f4269a3cc19205/loadtest/config.go#L27).

File: cmd/loadtest/loadtest.go [hunk 2]
```diff
@@ -31,10 +40,18 @@ func RunLoadTestCmdF(cmd *cobra.Command, args []string) error {
...
 		}
 		ue := userentity.New(memstore.New(), ueConfig)
-		return simplecontroller.New(id, ue, status)
+		switch controllerType {
+		case loadtest.UserControllerSimple:
+			return simplecontroller.New(id, ue, status)
+		case loadtest.UserControllerSimulative:
+			return simulcontroller.New(id, ue, status)
+		default:
+			// This should never happen if we validated the config.
```
Comment (isacikgoz) [inline cmd/loadtest/loadtest.go]: I see a good opportunity for adding a `panic()` here :)

File: loadtest/config.go [hunk 3]
```diff
@@ -40,6 +40,56 @@ func (cc *ConnectionConfiguration) IsValid() (bool, error) {
 	return true, nil
 }
 
+// UserControllerType describes the type of a UserController.
+type UserControllerType string
```
Comment (isacikgoz) [inline loadtest/config.go]: We can use unexported type here to avoid checks (L:59..62) below.

File: loadtest/config.go [hunk 4]
```diff
@@ -40,6 +40,56 @@ func (cc *ConnectionConfiguration) IsValid() (bool, error) {
...
+	}
+}
+
+// UserControllerConfiguration holds information about the UserController to
+// run during a load-test.
+type UserControllerConfiguration struct {
+	// The type of the UserController to run.
+	// Possible values:
+	//   "simple" - The simple version of a controller.
+	//   "simulative" - A more realistic controller.
```
Comment (isacikgoz) [inline loadtest/config.go]: We can replace these with constant types that we have.
```golang
UserControllerSimple     UserControllerType = "simple"
UserControllerSimulative                    = "simulative"
```

File: loadtest/config.go [hunk 5]
```diff
@@ -40,6 +40,56 @@ func (cc *ConnectionConfiguration) IsValid() (bool, error) {
...
+	// A rate multiplier that will affect the speed at which user actions are
+	// executed by the UserController.
+	// A Rate of < 1.0 will run actions at a faster pace.
+	// A Rate of 1.0 will run actions at the default pace.
+	// A Rate > 1.0 will run actions at a slower pace.
+	Rate float64
+}
+
+// IsValid reports whether a given UserControllerConfiguration is valid or not.
+func (ucc *UserControllerConfiguration) IsValid() (bool, error) {
```
Comment (isacikgoz) [inline loadtest/config.go]: Again, there is no case for `true, error` or `false ,nil`.

File: cmd/loadtest/loadtest.go (hunk 1 above)
Comment (streamer45) [reply cmd/loadtest/loadtest.go]: This pattern is actually shared by all config validation methods in this repo. I am wondering if @agnivade might have more info about this particular case.

If we decide to change I propose we do it in a separate PR that addresses all cases at once.

File: cmd/loadtest/loadtest.go (hunk 2 above)
Comment (streamer45) [reply cmd/loadtest/loadtest.go]: Sounds good.

File: cmd/loadtest/loadtest.go (hunk 1 above)
Comment (agnivade) [reply cmd/loadtest/loadtest.go]: I think I agree with @isacikgoz here. Almost all occurences of "isValid", "valid", "checkValid" in the standard library either return true/false if it does not need to distinguish between the type of error returned, or just error if it needs to distinguish.


One such example of the latter is: `crypto/x509/verify.go:func (c *Certificate) isValid`.

Let's do it in a separate PR.

File: loadtest/config.go [hunk 6]
```diff
@@ -40,6 +40,56 @@ func (cc *ConnectionConfiguration) IsValid() (bool, error) {
...
+		return false, fmt.Errorf("UserControllerType %s is not valid", t)
+	}
+}
+
+// UserControllerConfiguration holds information about the UserController to
+// run during a load-test.
+type UserControllerConfiguration struct {
+	// The type of the UserController to run.
+	// Possible values:
+	//   UserControllerSimple - The simple version of a controller.
```
Comment (agnivade) [inline loadtest/config.go]: A simple version ...

**Pull request merged**